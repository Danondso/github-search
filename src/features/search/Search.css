#search-container {
    display: flex;
    flex-direction: column;
    width: 100%;
}

#paging-container {
    padding: 1rem;
}

#total-results-container, #error-message-container {
    width: 100%;
    justify-content: center;
}

#page-number-container {
    padding: 0.5rem;
}

#paging-button-wrapper {
    display: flex;
    justify-content: space-evenly;
}

#result-layout-container {
    display: flex;
    justify-content: center;
}

/* lifted from https://css-tricks.com/centering-in-css/ */
.gentle-flex {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 1ch;
}

.loader-layout-container {
    padding: 1rem;
}

.result-layout-container {
    width: 30%;
}

a:link, a:hover {
    text-decoration: none;
}
