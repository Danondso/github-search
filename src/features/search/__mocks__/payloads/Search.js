export default {
  total_count: 3,
  incomplete_results: false,
  items: [
    {
      login: 'Danondso',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },
    {
      login: 'GerardWay',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },
    {
      login: 'RayToro',
      avatar_url: 'https://avatars.githubusercontent.com/u/7014871?v=4',
      url: 'https://api.github.com/users/Danondso',
      html_url: 'https://github.com/Danondso',
      repos_url: 'https://api.github.com/users/Danondso/repos',
    },

  ],
};
