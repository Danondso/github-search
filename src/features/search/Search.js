import React, { useEffect, useState } from 'react';
import axios from 'axios';
import {
  debounce, isEmpty,
} from 'lodash';
import { BallTriangle } from 'react-loader-spinner';
import Input from '../../components/Input/Input';
import {
  Card,
  CardHeader,
  CardContent,
} from '../../components/Card/Card';
import './Search.css';
import Avatar from '../../components/Avatar/Avatar';
import Pagination from '../../components/Pagination/Pagination';

import UserDetails from './UserDetails/UserDetails';

const authorizationHeader = {
  headers: {
    Authorization: `Basic ${btoa(`${process.env.REACT_APP_GITHUB_CLIENT_ID}:${process.env.REACT_APP_GITHUB_CLIENT_SECRET}`, 'base64')}`,
  },
};

function Search() {
  const [searchTerm, setSearchTerm] = useState('');
  const [errorMessage, setErrorMessage] = useState('');
  const [totalPages, setTotalPages] = useState(0);
  const [pageNumber, setPageNumber] = useState(1); // github's API defaults to 1
  const [searchResults, setSearchResults] = useState([]);
  const [totalResults, setTotalResults] = useState(0);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (!searchTerm) {
      setTotalResults(0);
      setTotalPages(0);
      setPageNumber(1);
      setSearchResults({});
      setErrorMessage('');
      return;
    }

    setIsLoading(true);

    axios.get(`https://api.github.com/search/users?q=${encodeURIComponent(searchTerm)}&page=${pageNumber}`, authorizationHeader).then((results) => {
      const { data } = results;
      const { total_count, items } = data;

      if (isEmpty(items)) {
        setErrorMessage('No Results Returned');
        setIsLoading(false);
        return;
      }

      setTotalResults(total_count);
      setTotalPages(Math.ceil(total_count / 30));
      setErrorMessage('');
      setSearchResults(items);
      setIsLoading(false);
    }).catch((error) => {
      setIsLoading(false);
      setSearchResults({});
      setErrorMessage(error.message || 'An error occurred.');
    });
  }, [searchTerm, pageNumber]);

  const handleOnSearch = (event) => setSearchTerm(event.target.value);
  const handlePaginate = (event) => {
    if (event.selected === 0) {
      setPageNumber(1);
    } else {
      setPageNumber(event.selected);
    }
  };

  const renderTotalResults = () => (totalResults > 0 && (
  <div id="total-results-container">
    {`Total Results: ${totalResults}`}
    <div id="page-number-container">
      {`Page ${pageNumber} of ${totalPages}`}
    </div>
  </div>
  ));

  const renderErrorMessage = () => (errorMessage && (
    <div id="error-message-container">
      {errorMessage}
    </div>
  ));

  const renderLoader = () => (
    <div className="gentle-flex loader-layout-container">
      <BallTriangle
        height="40"
        width="40"
        color="powderblue"
        ariaLabel="loading-indicator"
      />
    </div>
  );

  const renderSearchResults = () => (
    <div id="result-layout-container">
      <div className="gentle-flex result-layout-container">
        {searchResults.length > 0 && (searchResults.map((result) => (
          <Card key={result.login}>
            <a target="_blank" rel="noreferrer" href={result.html_url}>
              <CardHeader headerText={result.login} />
            </a>
            <CardContent>
              <Avatar imageSrc={result.avatar_url} altText={`Avatar of ${result.login}`} />
              <div className="user-details-wrapper">
                <UserDetails userName={result.login} />
              </div>
            </CardContent>
          </Card>
        )))}
      </div>
    </div>
  );

  const renderPaging = () => (
    <div id="paging-container">
      <div id="paging-button-wrapper">
        <Pagination handlePaginate={handlePaginate} totalPages={totalPages} />
      </div>
    </div>
  );

  return (
    <div id="search-container">
      <h2>Search Github Users</h2>
      <div id="input-container">
        <Input
          placeholder="Search For Users"
          onChange={debounce(handleOnSearch, 700)}
          autoFocus
        />
      </div>

      {renderPaging()}
      {renderTotalResults()}
      {renderErrorMessage()}
      {isLoading && renderLoader()}
      {renderSearchResults()}

      {/* As a user,
    ‚óè I see notable information for each search result, such as the description, star/follower
    count */}

    </div>
  );
}
export default Search;
