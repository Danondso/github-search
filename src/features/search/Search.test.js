import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import { act } from '@testing-library/react';
import Search from './Search';

import searchPayload from './__mocks__/payloads/Search';

jest.mock('axios');

describe('UserDetails Test Suite', () => {
  // the entire container setup to render this is lifted from https://reactjs.org/docs/testing-recipes.html#act
  let container;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  test('Renders correctly', async () => {
    axios.get.mockReturnValueOnce({ data: searchPayload });
    await act(async () => {
      ReactDOM.render(<Search />, container);
    });
    expect(container).toMatchSnapshot();
  });
});
