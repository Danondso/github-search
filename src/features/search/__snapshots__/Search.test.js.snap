// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Test Suite Renders correctly 1`] = `
<div>
  <div
    id="search-container"
  >
    <h2>
      Search Github Users
    </h2>
    <div
      id="input-container"
    >
      <input
        class="input"
        placeholder="Search For Users"
      />
    </div>
    <div
      id="paging-container"
    >
      <div
        id="paging-button-wrapper"
      />
    </div>
    
    <div
      id="result-layout-container"
    >
      <div
        class="gentle-flex result-layout-container"
      />
    </div>
  </div>
</div>
`;
